nifi-nar-bundles/nifi-ccda-bundle/nifi-ccda-processors/src/main/java/org/apache/nifi/processors/ccda/ExtractCCDAAttributes.java

```diff
@@ -0,0 +1,329 @@
+import org.openhealthtools.mdht.uml.cda.hitsp.HITSPPackage;
+import org.openhealthtools.mdht.uml.cda.ihe.IHEPackage;
+import org.openhealthtools.mdht.uml.cda.util.CDAUtil;
+import org.openhealthtools.mdht.uml.cda.util.CDAUtil.ValidationHandler;
+
+@SideEffectFree
+@SupportsBatching
+@InputRequirement(Requirement.INPUT_REQUIRED)
+@Tags({"CCDA", "healthcare", "extract", "attributes"})
+@CapabilityDescription("Extracts information from an Consolidated CDA formatted FlowFile and provides individual attributes "
+        + "as FlowFile attributes. The attributes are named as <Parent> <dot> <Key>. "
+        + "If the Parent is repeating, the naming will be <Parent> <underscore> <Parent Index> <dot> <Key>. "
+        + "For example, section.act_07.observation.name=Essential hypertension")
+public class ExtractCCDAAttributes extends AbstractProcessor {
+
+    private static final char FIELD_SEPARATOR = '@';
+    private static final char KEY_VALUE_SEPARATOR = '#';
+
+    private Map<String, Map<String, String>> processMap = new LinkedHashMap<String, Map<String, String>>(); // stores mapping data for Parser
+    private JexlEngine jexl = null; // JEXL Engine to execute code for mapping
+    private JexlContext jexlCtx = null; // JEXL Context to hold element being processed
+
+    private List<PropertyDescriptor> properties;
+    private Set<Relationship> relationships;
+
+    /**
+     * SKIP_VALIDATION - Indicates whether to validate the CDA document after loading.
+     * if true and the document is not valid, then ProcessException will be thrown
+     */
+    public static final PropertyDescriptor SKIP_VALIDATION = new PropertyDescriptor.Builder().name("skip-validation")
+            .displayName("Skip Validation").description("Whether or not to validate CDA message values").required(true)
+            .allowableValues("true", "false").defaultValue("true").addValidator(StandardValidators.BOOLEAN_VALIDATOR)
+            .build();
+
+    /**
+     * REL_SUCCESS - Value to be returned in case the processor succeeds
+     */
+    public static final Relationship REL_SUCCESS = new Relationship.Builder()
+            .name("success")
+            .description("A FlowFile is routed to this relationship if it is properly parsed as CDA and its content stored in FlowFile")
+            .build();
+
+    /**
+     * REL_FAILURE - Value to be returned in case the processor fails
+     */
+    public static final Relationship REL_FAILURE = new Relationship.Builder()
+            .name("failure")
+            .description("A FlowFile is routed to this relationship if it cannot be parse CDA and store content to FlowFile.")
+            .build();
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    protected void init(final ProcessorInitializationContext context) {

```

Comment:
Let's use an unmodifiableSet and unmodifiableList for the relationships and properties here.